lib/llvmopencl/WorkitemLoops.cc

```diff
@@ -1051,34 +1056,44 @@ void WorkitemLoopsImpl::addContextSaveRestore(llvm::Instruction *Def) {
       RematCandidate = false;
 
     if (StoreInst *ST = dyn_cast<StoreInst>(User)) {
-      if (!isa<UndefValue>(ST->getValueOperand())) {
-        Stores++;
-        if (Stores == 1) {
-          InitializerStore = ST;
-        } else {
-          InitializerStore = nullptr;
-          RematCandidate = false;
-#ifdef DEBUG_WORK_ITEM_LOOPS
-          std::cerr << "#### Multiple stores\n";
-          User->dump();
-#endif
-        }
-        if (PrevStoreRegion == nullptr) {
-          PrevStoreRegion = PRegion;
-        } else if (PrevStoreRegion != PRegion) {
-          RematCandidate = false;
-#ifdef DEBUG_WORK_ITEM_LOOPS
-          std::cerr << "#### Stores from multiple regions\n";
-          User->dump();
-#endif
-        }
-        if (LI.getLoopFor(ST->getParent()) != nullptr) {
-          RematCandidate = false;
-#ifdef DEBUG_WORK_ITEM_LOOPS
-          std::cerr << "#### Stores from inside a loop\n";
-          User->dump();
-#endif
-        }
+      // Stores of undefined values need not to be counted as actual stores
+      // since what we read from that location after that is undefined,
+      // thus could be as well the defined value of the another store.
+      if (isa<UndefValue>(ST->getValueOperand()))
+        continue;
+
+      Stores++;
+
+      // Another corner case to consider here is the case when an alloca
+      // is written inside a conditional block which could be even inside
+      // a diverging BB. The first intuition in that case would be to not
+      // allow rematerialization since we generally don't know at compile
+      // time if the branch is taken or not. However, when it comes to
+      // undefined behavior, reading from an uninitialized alloca is undefined,
```

Comment:
Seems to me ambiguous whether "undefined behavior" or "undefined value(s)" is meant here.

```suggestion
      // undefined behavior, reads from uninitialized allocas produce undefined values,
```